staker-rewards/lib/cache.js

```diff
@@ -0,0 +1,181 @@
+
+import { getPastEvents, callWithRetry } from "./contract-helper.js"
+import { KeepStatus, getKeepStatus } from "./keep.js"
+
+const DATA_DIR_PATH = path.resolve(process.env.DATA_DIR_PATH || "./data")
+const CACHE_PATH = path.resolve(DATA_DIR_PATH, "cache.json")
+
+// Our expectation on how deep can chain reorganization be. We need this
+// parameter because the previous cache refresh could store data that
+// are no longer valid due to a chain reorganization. To overcome this
+// problem we lookup `REORG_DEPTH_BLOCKS` earlier than the last refresh
+// block when starting the cache refresh.
+const REORG_DEPTH_BLOCKS = 12
+
+const CONCURRENCY_LIMIT = 3
+
+export default class Cache {
+  constructor(web3, contracts) {
+    this.web3 = web3
+    this.contracts = contracts
+  }
+
+  async initialize() {
+    this.cache = low(new FileSync(CACHE_PATH))
+    await this.cache
+        .defaults({
+          keeps: [],
+          lastRefreshBlock: this.contracts.factoryDeploymentBlock,
+        })
+        .write()
+  }
+
+  async refresh() {
+    const factory = await this.contracts.BondedECDSAKeepFactory.deployed()
+
+    const previousRefreshBlock = this.cache.get("lastRefreshBlock").value()
+
+    const startBlock =
+        previousRefreshBlock - REORG_DEPTH_BLOCKS > 0
+            ? previousRefreshBlock - REORG_DEPTH_BLOCKS
+            : 0
+
+    console.log(
+        `Looking for keeps created since block [${startBlock}]...`
+    )
+
+    const keepCreatedEvents = await getPastEvents(
+        this.web3,
+        factory,
+        "BondedECDSAKeepCreated",
+        startBlock
+    )
+
+    const newKeeps = []
+    keepCreatedEvents.forEach((event) => {
+      newKeeps.push({
+        address: event.returnValues.keepAddress,
+        members: event.returnValues.members,
+        creationBlock: event.blockNumber,
+      })
+    })
+
+    const cachedKeepsCount = this.cache.get("keeps").value().length
+
+    console.log(
+        `Number of keeps created since block ` +
+        `[${startBlock}]: ${newKeeps.length}`
+    )
+
+    console.log(`Number of keeps in the cache: ${cachedKeepsCount}`)
+
+    const actions = []
+    newKeeps.forEach((keep) => {
+      const currentlyCached = this.cache
+        .get("keeps")
+        .find({ address: keep.address })
+        .value()
+
+      if (!currentlyCached) {
+        actions.push(() => this.fetchKeep(keep))
+      }
+    })
+
+    if (actions.length === 0) {
+      console.log("Cached keeps list is up to date")
+    } else {
+      console.log(
+          `Fetching information about [${actions.length}] new keeps...`
+      )
+
+      const results = await pAll(actions, {
+        concurrency: CONCURRENCY_LIMIT,
+      })
+
+      console.log(
+          `Successfully fetched information about [${results.length}] new keeps`
+      )
+    }
+
+    const latestBlockNumber = keepCreatedEvents.slice(-1)[0].blockNumber
+    this.cache.assign({ lastRefreshBlock: latestBlockNumber }).write()
+
```

Comment:
Would it be possible to extract most of the logic above to `await this.fetchNewKeeps()`?  `refresh` function could be a little more elegant this way: `setup logic + fetchNewKeeps() + refreshActiveKeeps()`.